### PR TITLE
fixed MEF composition issue cased by waiter change

### PR DESCRIPTION
**Pull request by heejaechang**
### Customer scenario

User tries to use VB snippets and it is missing from snippet list.

### Bugs this fixes

https://github.com/dotnet/roslyn/issues/25212

### Workarounds, if any

There is no workaround

### Risk

There is no risk

### Performance impact

no performance impact

### Is this a regression from a previous update?

Yes. #24512 introduced the bug.

### Root cause analysis

when we changed waiter, `<ImportMany>` wasn't removed from signature causing MEF import exception to happen for this particular code. problem was that it didn't crash VS but silently logged the issue and moved on. so this wasn't discovered at the time.

### How was the bug found?

Dogfooding

### CODE REVIEW

**Comment by heejaechang** [comment]
@dotnet/roslyn-ide @davkean can you take a look? simple change.

File: src/VisualStudio/VisualBasic/Impl/Snippets/VisualBasicSnippetInfoService.vb [hunk 1]
```diff
@@ -18,7 +18,7 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.Snippets
         Inherits AbstractSnippetInfoService
 
         <ImportingConstructor>
-        Public Sub New(serviceProvider As SVsServiceProvider, <ImportMany> listenerProvider As IAsynchronousOperationListenerProvider)
+        Public Sub New(serviceProvider As SVsServiceProvider, listenerProvider As IAsynchronousOperationListenerProvider)
```
Comment (heejaechang) [inline src/VisualStudio/VisualBasic/Impl/Snippets/VisualBasicSnippetInfoService.vb]: this is actual change.

**Comment by jinujoseph** [comment]
So the set of deletes are just general cleanup and nothing to do with this fix ?

**Comment by davkean** [comment]
👍 On the ImportMany removal.

**Comment by heejaechang** [comment]
the ImportMany was due to Waiter change. so while I was on it, I cleaned up other code related to it as well.

**Comment by heejaechang** [comment]
ping?

File: src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs [hunk 2]
```diff
@@ -1,38 +0,0 @@
...
-using System;
-using System.Collections.Generic;
-using System.Linq;
-using Roslyn.Utilities;
-using System.Runtime.CompilerServices;
-
-namespace Microsoft.CodeAnalysis.Shared.TestHooks
-{
-    [Obsolete("use IAsynchronousOperationListenerProvider")]
-    internal class AggregateAsynchronousOperationListener : IAsynchronousOperationListener
```
Comment (jasonmalinowski) [inline src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs]: Wasn't this a shim you left for TypeScript to continue working? Did they move off of it now?

**Comment by jasonmalinowski** [comment]
So, um, tests. Do we not have any tests for snippet completion? Sounds like we need some here, pronto.

File: src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs (hunk 2 above)
Comment (heejaechang) [reply src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs]: yes, it was shims. forgot to tag them. but due to [Obsolete], they should have build break if they didn't move. also actual waiter was deleted way before. 

@amcasey

**Comment by heejaechang** [comment]
I dont know whether we had any tests over VB snippets that are affected by this MEF failure.

tagging @dpoeschl

File: src/VisualStudio/VisualBasic/Impl/Snippets/VisualBasicSnippetInfoService.vb (hunk 1 above)
Comment (jinujoseph) [reply src/VisualStudio/VisualBasic/Impl/Snippets/VisualBasicSnippetInfoService.vb]: 👍

File: src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs (hunk 2 above)
Comment (amcasey) [reply src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs]: FYI @uniqueiniquity

Comment (uniqueiniquity) [reply src/Features/Core/Portable/Shared/TestHooks/Legacy/AggregateAsynchronousOperationListener.cs]: Just moved, should be good to go.

**Comment by heejaechang** [comment]
I talked with @dpoeschl 

looks like we don't have vb snippet integration tests that cover this area. unlike C#, VB snippet still works without us providing this service. and VB doesn't include snippet in completionset so DDRIT won't catch it.

only way this affects is some tooltip text when snippet is invoked. and it looks like there is no test that cover that part.

**Comment by heejaechang** [comment]
opened issue - https://github.com/dotnet/roslyn/issues/25389 - to create new integration test

**Comment by heejaechang** [comment]
so do I need more code review for this PR? @Pilchie can I get approval for 15.7?

**Comment by sharwell** [comment]
:bulb: @heejaechang When writing comments on GitHub, you've noticed that the syntax `<ImportMany>` does not render in the comment (it's treated like an unsupported HTML tag that gets dropped). However, rather than rewrite it as `< ImportMany >`, you can rewrite it as `` `<ImportMany>` `` to get better overall display.

| Markdown | Rendered | Verdict |
| --- | --- | --- |
| `<ImportMany>` | <ImportMany> | 😕 |
| `< ImportMany >` | < ImportMany > | 😑 |
| `` `<ImportMany>` `` | `<ImportMany>` | 😀 |

**Comment by jinujoseph** [comment]
@Pilchie for 15.7.P2 ask mode approval

**Comment by Pilchie** [comment]
Approved - thanks!

**Comment by Pilchie** [comment]
Actually wait - P2 is now in escrow.  If this is needed for P2, it needs a VSTS bug and approval (and possibly a different branch).

**Comment by heejaechang** [comment]
@jinujoseph where do you want this to be in? there is no missing features user can see from this. vb snippet still working. the thing missing is a word in the snippet tooltip. unlike C#, editor takes care of vb snippet and it doesn't show up in completion set.

tagging @Pilchie and @dpoeschl

**Comment by dpoeschl** [comment]
The only other thing this snippet info service does is populate the list when you do e.g. `m?<tab>` in VB, but I'd like to believe nobody uses that. @Pilchie do you know of people wanting that in modern times (I have this question for a couple reasons, including other potential changes to completion).

Regardless, I'm okay with pushing this past P2 (the actual snippet fix, don't know about the obsolete API removal).

**Comment by Pilchie** [comment]
@dpoeschl I've never specifically heard of anyone doing that ever :)  OTOH - do we have telemetry to know how often it's used?

**Comment by jinujoseph** [comment]
@Pilchie @heejaechang , lets take this for 15.7.P3 ( i don't have a particular reason to push for P2 )

**Comment by Pilchie** [comment]
Okay - sounds good.  Approved for 15.7 P3.

**Pull request merged**